Spark 3.4.0 and DBR 12.2 support

chris-twiner opened the pull request:
#698 - until 3.4.0 is released it won't compile

Closes #698
Closes #506

chris-twiner commented:
fyi - I've tested with #700 as well on the final cut, would you like me to roll #700 (doCodegen instead of codegen) into this PR?

pomadchin commented:
@chris-twiner could you rebase on top of main branch? (I can do it too, but it may lead to issues on your machine)

pomadchin commented:
Fixed builds here 🤦 https://github.com/typelevel/frameless/pull/709

chris-twiner commented:
based on coverage results the scalareflection stuff in FramelessInternals is trimmed down to a bare minimum.  I have also removed the column (_.expr) function as it's not used anywhere.

The test failure [here](https://github.com/chris-twiner/frameless/actions/runs/4898136130/jobs/8746945869#step:9:844) I've also fixed for rlike to check for a valid generated regex.

chris-twiner commented:
Just saw that 0.14 went out, which explains a lot, I'll correct mima stuffs and the readme in a few hours

chris-twiner commented:
test fail of [avg](https://github.com/chris-twiner/frameless/actions/runs/4902770544/jobs/8754811976) due to overflow in AggregateFunctionsTests:127 avg check(forAll(prop[Long, Double] _)), 3.4 moved that out of the normal checks per 39316.  A restart completed but this also should be limited to non-overflow cases, just not sure how easily to do that right now.

chris-twiner commented:
wrt KMeansTest failure, the euclideanUpdaeInPlace ArrayIndexOutOfBoundsException now occurs in 45 out of 1000 runs it seems (i.e. out of 10,000 maximum actual checks ~0.45 -> 4.5% of the time).

Running the average tests, similarly failure didn't occur in 500(0) runs (so prob far less than 0.01% occurrence).  The actual input was described in the exception as a List of strings rather than a List of Longs, which is odd, I cannot reproduce.

I don't know that it makes sense to try and resolve these tests further given the low percentage occurrence *and* that the "issues" are not in frameless code.  As such I've added a wrapper to retry in the case of that specific issue for kmeans and avg.  Alternatively the [spark test approach](https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala#L497) could be used instead of scalacheck for kmeans, just consting the vectors.

chris-twiner commented:
as review hasn't started yet I hope this approach is ok - via imports only, there was too much code change for something that should be transparent. 

Note the two lines of code shown as untested in previous patches (code gen for DisambiguiateLeft/Right) were never tested, I've forced the issue.

chris-twiner commented:
@pomadchin  - sorry I just realised you may not have been alerted it was green - hopefully this does alert you, since there are other releases in between the mima stuff may need version bumps though.

pomadchin commented:
@chris-twiner yes!! Thanks for pointing it out; I’ll hopefully take a closer look until this Friday, or the beginning of the next week.

chris-twiner commented:
np, fyi - 3.5.0 doesn't seem to have any issues with this [snapshot](https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-core_2.12/3.5.0-SNAPSHOT/spark-core_2.12-3.5.0-20230528.001246-122.jar).

pomadchin commented on ml/src/test/scala/frameless/ml/clustering/KMeansTests.scala:
```diff
@@ -12,37 +12,54 @@ import frameless.ml.params.kmeans.KMeansInitMode
 
-    check(prop)
-    check(prop3[Double])
+    tolerantRun( _.isInstanceOf[ArrayIndexOutOfBoundsException] ) {
```
Do we still need it?

chris-twiner replied:
without the tolerantRun it [fails often enough](https://github.com/typelevel/frameless/pull/699#issuecomment-1538614609) on 3.4 to stop the build

pomadchin reviewed (approved):
LGTM! A nice approach. I cleaned up the build file a bit but rather than that the change looks good. Thanks for the incredible work!

pomadchin commented:
Created a follow up issue https://github.com/typelevel/frameless/issues/717

pomadchin commented:
Thanks for a great contribution one more time! 🔥

chris-twiner commented:
> Thanks for a great contribution one more time! 🔥

As with [the original 3.x](https://github.com/typelevel/frameless/pull/433) you are mtw!
